pkg/api/alerting/v2alpha1/types_test.go

```diff
@@ -0,0 +1,360 @@
+		Annotations: map[string]string{
+			"alias":       "The alias is here",
+			"description": "The description is here",
+		},
+		Labels: map[string]string{
+			"alias":       "The alias is here",
+			"description": "The description is here",
+		},
+		ActiveAt: &time.Time{},
+	}
+
+	alerts = []*Alert{
+		alert,
+		{
+			RuleId:   "ca7f09e76954e699",
+			RuleName: "test-cpu-usage-high-2",
+			State:    rules.StateFiring.String(),
+			Value:    `namespace:workload_cpu_usage:sum{namespace="test"} > 1`,
+			Annotations: map[string]string{
+				"alias":       "The alias is here",
+				"description": "The description is here",
+			},
+			Labels: map[string]string{
+				"alias":       "The alias is here",
+				"description": "The description is here",
+			},
+			ActiveAt: &time.Time{},
+		},
+	}
+)
+
+func TestAlertingRuleQueryParams_Filter(t *testing.T) {
+	// empty
+	// Alert Array is empty
+	loacalAlerts := []*Alert{}
+	queryParam := &AlertQueryParams{}
+	ret := queryParam.Filter(loacalAlerts)
+	assert.Empty(t, ret)
+
+	// Alert Array has nil
+	loacalAlerts = []*Alert{nil}
+	ret = queryParam.Filter(loacalAlerts)
+	assert.Empty(t, ret)
+
+	// all pass
+	loacalAlerts = alerts
+	ret = queryParam.Filter(loacalAlerts)
+	assert.NotEmpty(t, ret)
+}
+
+func TestAlertingRuleQueryParams_matches(t *testing.T) {
+	loacalAlert := alert
+	queryParam := alertQueryParams
+	queryParam.LabelEqualFilters = map[string]string{
+		"flags":   "The flags is here",
+		"cpu_num": "The cpu_num is here",
+	}
+	ret := queryParam.matches(loacalAlert)
+	assert.False(t, ret)
+
+	queryParam.LabelEqualFilters = map[string]string{
+		"alias":       "The alias is here",
+		"description": "The description is here",
+	}
+	ret = queryParam.matches(loacalAlert)
+	assert.True(t, ret)
+}
+
+func TestAlertingRuleQueryParams_Sort(t *testing.T) {
+	loacalAlerts := alerts
+	queryParam := alertQueryParams
+	queryParam.Sort(loacalAlerts)
+}
+
+func TestAlertingRuleQueryParams_Sub(t *testing.T) {
+	loacalAlerts := alerts
+	queryParam := alertQueryParams
+	queryParam.Sub(loacalAlerts)
+}
+
+var (
+	succBulkItemResponse = BulkItemResponse{
+		RuleName:  "test-cpu-usage-high",
+		Status:    StatusSuccess,
+		Result:    ResultCreated,
+		ErrorType: "",
+		Error:     nil,
+		ErrorStr:  "",
+	}
+
+	errBulkItemResponse = BulkItemResponse{
+		RuleName:  "test-mem-usage-high",
+		Status:    StatusError,
+		Result:    ResultUpdated,
+		ErrorType: ErrBadData,
+		Error:     errors.New(string(ErrBadData)),
+		ErrorStr:  string(ErrBadData),
+	}
+
+	bulkResponse = BulkResponse{
+		Items: []*BulkItemResponse{
+			&succBulkItemResponse,
+			&errBulkItemResponse,
+		},
+	}
+)
+
+func TestParseAlertingRuleQueryParams(t *testing.T) {
+	queryParam := "/alerting.kubesphere.io/v2alpha1/rules?name=test-cpu&state=firing&health=ok&sort_field=lastEvaluation&sort_type=desc&label_filters=name~test"
+	req, err := http.NewRequest("GET", fmt.Sprintf("http://localhost?%s", queryParam), nil)
```

Comment:
Is this a redundant `?`?

Reply:
Obtain the http.Request object required by the restful.NewRequest function through the http.NewRequest function (various RULs) . The restful.NewRequest function returns restful.Request , which is the parameter required by the ParseQueryParameter function.  it's necessary.